From e8d62a314805b41df121e0ec3d3246bb69f7a508 Mon Sep 17 00:00:00 2001
From: zhu yixin <yixin.zhu@intel.com>
Date: Mon, 16 Jul 2018 14:54:41 +0800
Subject: [PATCH] Add CAP_SYS_ADMIN check on export and unexport.
 export/unexport by default only root can operate. The extra sanity check is
 based on Security SE's assumption: Root can alter the sysfs's permission so
 that other non-root user may be able to operate this sysfs. Extra check here
 to guarantee only root can do this change.

---
 drivers/gpio/gpiolib-sysfs.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/gpio/gpiolib-sysfs.c b/drivers/gpio/gpiolib-sysfs.c
index 4b44dd97c07f..43fa12a75c05 100644
--- a/drivers/gpio/gpiolib-sysfs.c
+++ b/drivers/gpio/gpiolib-sysfs.c
@@ -446,6 +446,9 @@ static ssize_t export_store(struct class *class,
 	struct gpio_desc	*desc;
 	int			status;
 
+	if (!capable(CAP_SYS_ADMIN))
+		return -EPERM;
+
 	status = kstrtol(buf, 0, &gpio);
 	if (status < 0)
 		goto done;
@@ -488,6 +491,9 @@ static ssize_t unexport_store(struct class *class,
 	struct gpio_desc	*desc;
 	int			status;
 
+	if (!capable(CAP_SYS_ADMIN))
+		return -EPERM;
+
 	status = kstrtol(buf, 0, &gpio);
 	if (status < 0)
 		goto done;
